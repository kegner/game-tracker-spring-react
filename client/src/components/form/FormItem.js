import React, { useState } from 'react'
import _ from 'lodash';

const FormItem = (props) => {

  const { id, type, label, value, onChange, currency, validClass, validClassSetter, isRequired } = props;

  const [inputId] = useState(_.uniqueId(id + "-"));

  let changeHandler = onChange;

  if (currency) {
    changeHandler = (e) => {
      // supports up to 999.99
      if (!e.target.value.match(/(?=.*)^(\d{1,3})?(\.\d{1,2})?$/)) {
        e.preventDefault();
        return false;
      }
      onChange(e);
    }
  }

  const blurHandler = () => {
    if (validClassSetter) {
      validClassSetter(!value ? "is-invalid" : "");
    }
  }
  
  return (
    <>
      <label htmlFor={inputId} className="form-label">
        {isRequired ? <span className="text-danger" aria-required="true">* </span> : ""}{label}
      </label>
      <input type={type} className={`form-control ${validClass}`} id={inputId} value={value}
        onChange={changeHandler} onBlur={blurHandler} />
    </>
  )
}

export default FormItem
