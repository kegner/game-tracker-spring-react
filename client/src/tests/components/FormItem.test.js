import React from 'react';
import { render, getByLabelText, fireEvent } from '@testing-library/react';
import FormItem from '../../components/form/FormItem';

let hasChanged = false;
const changeHandler = (e) => { hasChanged = true }

let component;
let container;
let asFragment;

describe("FormItem", () => {

  beforeEach(() => {
    hasChanged = false;
    component = render(
      <FormItem id="title" label="Title" value="My title" type="text"
          onChange={changeHandler} />
    );
    container = component.container;
    asFragment = component.asFragment;
  });

  it("exists", () => {
    expect(asFragment()).toMatchSnapshot();
  });

  it("shows values", () => {
    expect(getByLabelText(container, "Title")).toBeVisible();
  });

  it("change handler", () => {
    const input = container.querySelector("input");
    expect(hasChanged).toBe(false);
    fireEvent.change(input, { target: { value: "new title" } });
    expect(hasChanged).toBe(true);
  });

  it("change handler for currency", () => {
    component = render(
      <FormItem id="title" label="Title" value="My title" type="number"
          onChange={changeHandler} currency={true} />
    );
    container = component.container;
    asFragment = component.asFragment;

    const input = container.querySelector("input");
    expect(hasChanged).toBe(false);
    fireEvent.change(input, { target: { value: "1.123" } });
    expect(hasChanged).toBe(false);
  });
});