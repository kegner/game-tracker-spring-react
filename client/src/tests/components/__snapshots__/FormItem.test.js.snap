// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormItem exists 1`] = `
<DocumentFragment>
  <label
    class="form-label"
    for="title-1"
  >
    Title
  </label>
  <input
    class="form-control undefined"
    id="title-1"
    type="text"
    value="My title"
  />
</DocumentFragment>
`;
